domain/storage/service/storagepool.go

```diff
@@ -0,0 +1,246 @@
+	}
+	if providerType == "" {
+		return storageerrors.MissingPoolTypeError
+	}
+	if s.registry == nil {
+		return errors.Errorf("cannot validate storage provider config for %q without a registry", name)
+	}
+
+	cfg, err := storage.NewConfig(name, providerType, attrs)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	p, err := s.registry.StorageProvider(providerType)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := p.ValidateConfig(cfg); err != nil {
+		return errors.Annotate(err, "validating storage provider config")
+	}
+	return nil
+}
+
+// DeleteStoragePool deletes a storage pool, returning an error satisfying
+// [errors.NotFound] if it doesn't exist.
+func (s *StoragePoolService) DeleteStoragePool(ctx context.Context, name string) error {
+	// TODO(storage) - check in use when we have storage in dqlite
+	// Below is the code from state that will need to be ported.
+	/*
+		var inUse bool
+		cfg, err := sb.config(context.Background())
+		if err != nil {
+			return errors.Trace(err)
+		}
+		operatorStorage, ok := cfg.AllAttrs()[k8sconstants.OperatorStorageKey]
+		if sb.modelType == ModelTypeCAAS && ok && operatorStorage == poolName {
+			apps, err := sb.allApplications()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			inUse = len(apps) > 0
+		} else {
+			query := bson.D{{"constraints.pool", bson.D{{"$eq", poolName}}}}
+			pools, err := storageCollection.Find(query).Count()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			inUse = pools > 0
+		}
+		if inUse {
+			return errors.Errorf("storage pool %q in use", poolName)
+		}
+	*/
+	err := s.st.DeleteStoragePool(ctx, name)
+	return errors.Annotatef(err, "deleting storage pool %q", name)
+}
+
+// ReplaceStoragePool replaces an existing storage pool, returning an error
+// satisfying [storageerrors.PoolNotFoundError] if a pool with the name does not exist.
+func (s *StoragePoolService) ReplaceStoragePool(ctx context.Context, name string, providerType storage.ProviderType, attrs PoolAttrs) error {
+	// Use the existing provider type unless explicitly overwritten.
+	if providerType == "" {
+		existingConfig, err := s.st.GetStoragePoolByName(ctx, name)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		providerType = storage.ProviderType(existingConfig.Provider)
+	}
+
+	err := s.validateConfig(name, providerType, attrs)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	attrsToSave := transform.Map(attrs, func(k string, v any) (string, string) { return k, fmt.Sprint(v) })
+	sp := domainstorage.StoragePoolDetails{
+		Name:     name,
+		Provider: string(providerType),
+		Attrs:    attrsToSave,
+	}
+	err = s.st.ReplaceStoragePool(ctx, sp)
+	return errors.Annotatef(err, "replacing storage pool %q", name)
+}
+
+// ListStoragePools returns the storage pools matching the specified filter.
+func (s *StoragePoolService) ListStoragePools(ctx context.Context, filter domainstorage.StoragePoolFilter) ([]*storage.Config, error) {
```

Comment:
We should step away from `StoragePoolFilter` or `Filters` in general. I'd rather see more methods that target specific logic.

Reply:
But this comes from the cli. We can't just throw away filers for the sake of it.

Reply:
See my comment on the implementation above.